builder/builder.go

```diff
@@ -643,6 +656,43 @@ func (b *Builder) AddBundles(bundles []string, alllocal bool, allupstream bool,
 	return nil
 }
 
+// ListBundles prints out the bundles in the mix, all bundles in the mix
+// including recursive includes, all bundles available in local bundles, or
+// all bundles available in upstream bundles
+func (b *Builder) ListBundles(full bool, local bool, upstream bool) error {
+	var bundles []string
+	var err error
+
+	switch {
```

Comment:
This feels wrong. If a bundle can be at most full, local or upstream, then we should have a type saying so, rather than a set of bool. If on the other-hand it can be both full and upstream then we need a comment to say why being upstream is more important than being full.

Reply:
I think you're misunderstanding the conditions. A _bundles_ isn't "full" or "upstream", the _list_ is. You're either getting a "full list" or an "upstream list".

If you look at the flag's documentation, the command either lists the items in the mix bundles list, the _full_ set of bundles that will be included (by recursively unpacking the includes in the mix bundles list), all of the available local bundles, or all of the available upstream bundles.

You're right that the precedence of how the command behaves (i.e., passing `--upstream` trumps `--full`) needs to be documented.

Reply:
It is entirely possible I am misunderstanding, I am looking at the code rather than the big picture. As I said it "feels wrong". So let me rephrase the question - what are the valid states of the 3 bool values? The 3 bools look to just be to make it trivial to use Cobra, without any warning to the user that they are misunderstanding what they are asking for.

Reply:
Optional: maybe the feeling is an indication that the interface could be improved. Two possible venues here:

(a) you have 4 mutual exclusive cases, so make this visible in the code. Make your list function take an enum or give 4 different functions to the callsite. Callsite will handle the boolean drama matching.

(b) Rework the CLI.
- mixer bundle list: show all bundles in the bundle list, their dependencies (marked in some way) AND the local bundles NOT included (marked in some way).
- mixer bundle list --upstream: show the all upstream bundles, with those used/included by the bundle list marked.

And maybe in that case Builder would expose 2 different functions, instead one with a boolean.